lib/server.js

```diff
@@ -1264,41 +1267,149 @@ WalletService.prototype._checkTxAndEstimateFee = function(txp) {
       res[val] = val;
       return res;
     }, {});
 
-    utxos = _.reject(utxos, function(utxo) {
-      return excludeIndex[utxo.txid + ":" + utxo.vout];
+    return _.filter(utxos, function(utxo) {
+      if (utxo.locked) return false;
+      if (utxo.satoshis <= feePerInput) return false;
+      if (txp.excludeUnconfirmedUtxos && !utxo.confirmations) return false;
+      if (excludeIndex[utxo.txid + ":" + utxo.vout]) return false;
+      return true;
+    });
+  };
+
+  function partitionUtxos(utxos) {
+    return _.groupBy(utxos, function(utxo) {
+      if (utxo.confirmations == 0) return '0'
+      if (utxo.confirmations < 6) return '<6';
+      return '6+';
+    });
+  };
+
+  function select(utxos, cb) {
+    var totalValueInUtxos = _.sum(utxos, 'satoshis');
+    var netValueInUtxos = totalValueInUtxos - baseTxpFee - (utxos.length * feePerInput);
+
+    if (totalValueInUtxos < txpAmount) {
+      log.debug('Total value in all utxos (' + Utils.formatAmountInBtc(totalValueInUtxos) + ') is insufficient to cover for txp amount (' + Utils.formatAmountInBtc(txpAmount) + ')');
+      return cb(Errors.INSUFFICIENT_FUNDS);
+    }
+    if (netValueInUtxos < txpAmount) {
+      log.debug('Value after fees in all utxos (' + Utils.formatAmountInBtc(netValueInUtxos) + ') is insufficient to cover for txp amount (' + Utils.formatAmountInBtc(txpAmount) + ')');
+      return cb(Errors.INSUFFICIENT_FUNDS_FOR_FEE);
+    }
+
+    var bigInputThreshold = txpAmount * Defaults.UTXO_SELECTION_MAX_SINGLE_UTXO_FACTOR + (baseTxpFee + feePerInput);
+    log.debug('Big input threshold ' + Utils.formatAmountInBtc(bigInputThreshold));
+
+    var partitions = _.partition(utxos, function(utxo) {
+      return utxo.satoshis > bigInputThreshold;
+    });
+
+    var bigInputs = _.sortBy(partitions[0], 'satoshis');
+    var smallInputs = _.sortBy(partitions[1], function(utxo) {
+      return -utxo.satoshis;
+    });
+
+    log.debug('Considering ' + bigInputs.length + ' big inputs (' + Utils.formatUtxos(bigInputs) + ')');
+    log.debug('Considering ' + smallInputs.length + ' small inputs (' + Utils.formatUtxos(smallInputs) + ')');
+
+    var total = 0;
+    var selected = [];
+    var error;
+
+    _.each(smallInputs, function(input, i) {
+      log.debug('Input #' + i + ': ' + Utils.formatUtxos(input));
+
+      if (input.satoshis < feePerInput) {
+        log.debug('The input does not cover the extra fees (' + Utils.formatAmountInBtc(feePerInput) + ')');
+        return false;
+      }
+
+      var inputAmount = input.satoshis - feePerInput;
+      log.debug('The input contributes ' + Utils.formatAmountInBtc(inputAmount));
+
+      selected.push(input);
+
+      var txpSize = baseTxpSize + selected.length * sizePerInput;
+      var txpFee = baseTxpFee + selected.length * feePerInput;
+
+      log.debug('Tx size: ' + Utils.formatSize(txpSize) + ', Tx fee: ' + Utils.formatAmountInBtc(txpFee));
+
+      var amountVsFeeRatio = txpFee / txpAmount;
```

Comment:
You are right on both variable names. I just couldn't decide which ordering would be a more intuitive ratio so I changed it several times.